### ISSUE TITLE
[Suggested Feature] AOBD could provide an excellent basis for a materials API

### ISSUE THREAD

**Issue by kotoroshinoto** [MinecraftForge/MinecraftForge#1753]
ganymedes01 has agreed that it has strong potential to do so. He thinks it would probably be better for such a system to be part of the forge API than being a separate mod if you are willing to ask him to implement it and have it integrated.

Idea would be to provide modders the ability to request from the API, specific metals and metal related items/blocks/fluids.

A metal would be registered with a color, then from that metal's name you could provide the different types of crystals/clumps/dusts/nuggets/blocks/molten fluids/ore/ingots either from a decolorized set of default resource images that are colored when the class is instantiated, or from resources provided by the mods if they are putting in an item that doesn't HAVE a default resource to match (or can't be properly depicted by recoloring the default one). 

There is no inherent reason to have to completely replace the current oredict with this, as that provides ways to recognize other items that aren't necessarily ores as being identical. Centralizing the metals and metal materials would allow the potential to unify the metals being spawned during worldgen of various mods, and eliminate niggling issues of not being able to stack materials from different mods without passing them through an item, block, or machine that uses the oredict to convert them. (and have them be inherently registered into the oredict system in whatever form it takes as forge develops going forward)

**Comment by ganymedes01** [comment]
A modder would be able to register "copper" as a material and then forge would return it items, blocks and fluids (which would be generated and registered automatically as if by that mod). If another mod register "copper" it would get the already generated items out. Basically unifying every ingot, nugget, plate, dust, ore, block and etc of that specified type.

Users could use resource packs to change the appearance of those items (since it would "default" to whatever appearance the first mod to register it gave it).

**Comment by LexManos** [comment]
Forge does not add any in-game content.
Last time this was attempted it turned into modders bitching that they wanted to control everything and overwriting other's registrations.
The 'Ore' in 'OreDictionary' is just a legacy name, I'm not sure why people get stuck on that, it's a generic ItemStack equivalency system.

**Comment by kotoroshinoto** [comment]
I suppose that is fair. The idea here is that it would only generate items on request,and wouldn't seek to do anything other than color in a generic item. Recipes and the like would have to be handled by the mod authors themselves.

Ultimately such fights between modders should be rendered petty if the expectation is that ingots/blocks etc aren't to be allowed to implement extra methods themselves and all mod specific behaviors are to be handled by the mods internally by recognizing the item. 

They might still complain about color choices or initialized image resources. It seems something like that wouldn't be the end of the world, as they could provide resource pack components to users to help them make things look the way they want.

If you think such a thing is better off as a separate mod then maybe thats how it should be done.

**Comment by TheRoBrit** [comment]
I feel that while the premise for the idea is not by any means a bad one (unifying content to use some kind of standard). However, the idea of adding something like this to Forge is most definitely a bad idea. As Lex previously stated, Forge does not add any content and it tries to be as abstracted as possible. The whole notion of solely adding a system for "metals" does seem very flimsy on a surface-level.

On the practical implementation end, this would also require mod developers to register their textures in the same way (a generic grey-scaled version of the sprite/sprites) and then apply a render pass to provide their items/blocks with a particular colour. Which is both a waste of an additional render-pass on ALL items/blocks using this system and ultimately would not unify in a way many mod developers would be happy with. Some mod developers choose not to make their metals compatible/inter-changeable with others of the same name (by not adding to Ore Dictionary) and they also may wish not to have their 'metals' (or on a more abstracted level - items/blocks) to use the same colouration as another mod has registered previously - thus defeating the purpose)

tl;dr - The idea is a great way to standardize YOUR (as a mod developer) assets and implementation of these items/blocks, but not the assets and implementation of all mod developers wishing to add similar items/blocks.

**Comment by LexManos** [comment]
Yes, it can be done in a way that doesn't cause in-game content to be created unless requested by a mod.
And this is something that I WISH I could add to Forge as it would make life a lot simpler for modders and end users.
But the fact of the matter is that this, much like the power system, will never go in Forge due to the fact that modders can't agree on anything.
Take for example.
One mod adds Copper, We generate the ores and everything they behave just like Iron in game.
But what if another modder wanted it to behave like Redstone in the way that it drops rand# of dust.
Or What if one modder added lead, one modder wants it to be a normal ore, the other wants you to take damage when you touch it.
Its things like this that modders can never agree on and Forge is not the place to fight the battle.
We already have enough issues with modders hacking the Fluid registry so that there implementation of fluids are the ones used.

Don't know, may be feasible to go into forge with a massive IF THIS BREAKS ITS BECAUSE MODDERS.

**Comment by kotoroshinoto** [comment]
You could leave the handling of worldgen to the mods that request the item if you think that would be better. The mods would then do their own oregen passes, they'd just be putting the same copper in 2 places rather than 2 different coppers ;p.

knowing some modders they might register something as "MyCopper" instead of just "Copper" and fix the oredict entries themselves, but you can't force convention, its a bit like herding cats.

**Comment by laz2727** [comment]
Note that what you are talking about is pretty much a fully configurable
mix of aobd and tf.

**Comment by kotoroshinoto** [comment]
I obviously know what AOBD is. What is TF?

**Comment by laz2727** [comment]
Thermal Foundation.

**Comment by kotoroshinoto** [comment]
Ahh, I know what THAT is, the abbreviation didn't click XD.

I suppose that it would be very much like that.

**Comment by ganymedes01** [comment]
Most mods tend to handle the same ores in very similar fashion. If you look at the most common ones (copper, lead, silver, etc) they all behave in pretty much a similar way across the board. If any mod decided they want to implement copper that does something different then they'll just have to abstain from using forge's system. Would be worth it because it's a system that would fit well in 99% of cases, the 1% will just have to keep doing what they already do and go against the flow (after all, lots of mods add copper, if there's a mod out there that adds copper that behaves differently, compatibility is already an issue)

**Comment by phase** [comment]
I think @LexManos has a good point. There's no need for a system like this since no one can agree on anything. It's amazing how many different coppers there can be in a mod pack, each one doing different things. You *could* do this, but it'd be pointless.

**Comment by chibill** [comment]
I would also agree with lex its just like the ORE dict it is good but half the time the moders don't use right.

**Comment by kotoroshinoto** [comment]
it doesn't have to have complete adoption by all mods to be a useful feature for those that do choose to adopt it.

If you let modders handle the oregen passes themselves rather than by this system so there is no conflict in that area. the only complaints remaining would be arguments over how the items should look, and ultimately that can be solved by handing users files to put into a resourcepack so they can choose. 

If they REALLY feel the need to add additional functionality to the item/block class they can just forgo the use of this system.

**Comment by phase** [comment]
@kotoroshinoto Either no one's going to use it or everyone's going to use it. It can't and won't be in between.

**Comment by kotoroshinoto** [comment]
strange that you feel that way, not everybody uses the oredict stack identity system, why would this HAVE to be a mandatory system?

**Comment by KingLemming** [comment]
There's not really a great way to handle this for the reasons Lex has pointed out. Ensuring recipe cross-compat is what the OreDict does.

What you are asking for is actually removing options from server administrators. They may want multiple versions of a Copper Ore; there may be legitimate reasons for that.

**Comment by kotoroshinoto** [comment]
nowhere in the idea did we say FORCE the modders to use the system. If they wanted to say, create malachite ore AND native copper ore, that could be done. If this system could not support it, they could register the native copper with this system if they wanted, and then just add malachite on the side, or simply NOT use the system at all.

**Comment by HenryLoenwind** [comment]
Yeah, much potential for conflict if doing it this way.

However, what I'd love to have is a system where I can conditionally register my items and blocks. So I can tell it "if there's no ingotCopper in the oriDict, use this item." And it will make sure that if a world is loaded where another mod's ingotCopper was picked (e.g. because the modlist changed, or someone changed their after/before, or a mod added copper), it still works. (Obviously if the item that was picked for that world doesn't exist anymore, the item needs to be removed from the world.)

At the moment, only registering an item/block with the game if there's no oreDict for it, is a very bad idea. This is what leads to the multitude of different but oreDicted items in the game. Not that every mod author wants do do something different. If I want to use copper in a recipe, I must either add a dependency on another mod that has copper, or create another copper dupe in the game.

To come back to AOBD, a second call "here's an item for groundedCopperGranules, but don't register it unless another mod registers /.*Copper/" would be a nice addition.

**Comment by Spartan322** [comment]
The only way this could work is for people to accept to create a baseline, and follow the structure a bit. 90% of modders act like they hate that idea, and the bigger mods and systems try to make addons to do this, and while some succeed, every time, you get a larger group who break the curve. So simply and unfortunately, this either needs to be forced, which is asking for trouble, or it shouldn't happen,  not until the modding community gets some type of paradigm (which, so far, has never happened, and probably won't)

**Comment by cosmicdan** [comment]
While I can perfectly understand Lex's apathy in implementing such a system (some of the bickering that goes on with some modders watning to be the ore masters is pretty silly), I think it's a very possible thing - albeit with some changes that some might object to for no rational reason.

I don't agree with the point that a Resources API (I mean "resources" as in the generic term for an equivalency-enabled resource registration API) does not belong in Forge. Forge is an API, and API's should provide controlled implementations for creating content, and it's my opinion that the whole ore/resource generation and ore dictionary system as it is "too liberal" and has really outgrown itself. I have a lot of ideas on how to improve this and have had them for some time, wanting to do a mod for this resource-consolidation kind of thing, but even if I had the time and skill it's probably not feasible. Even with a buttload of ASM injections, saving the need to walk through the entire ore dictionary/registry and the recipe list, there are the issues of (a) all the non-standard recipes e.g. custom crafting tables and machine outputs, and (b) insufficient information associated with these resources/ores. We have MineTweaker for this kind of thing already, all this is technically doable with it, but it requires a lot of user effort and still has it's limitations - because MineTweaker is "only" a mod that has little info to work with.

Ahem... on to my actual thoughts.

I think a new (or enhanced, rather) system shouldn't be forced, but VERY encouraged - and possibly phase out the old system in future. HOWEVER, about the AOBD-like method, I agree that Forge itself should *not* have any additional assets such as dust/nuggets/poops/whatever-the-heck-else which act as grey-scaled templates for other mods to match against and color; but rather it needs to unify mod-added resources in a more coherent and descriptive database - something a bit better than the simple keyword matching of the oredict (which would still be needed, though, for consolidation processing). This could be done by requiring that all mod-added ore-dictionary-like items form a bidirectional, hierarchical relationship between the ores and it's refined products. Allow me to give some examples, based on Lex's mentioned issues:

1) Mod A adds a plain-jane copper ore. The associated ingotCopper would have a "parent" property which would be set to it's ore block, and likewise the ore block would have a "child" property which is set to ingotCopper. In this instance, the child/parent properties are easily set by Forge itself simply by the fact that the ore smelts directly into the ingot - all vanilla stuff. Forge will take care of that. So we have a simple relationship with an ore and it's product now, which Forge is aware of. Note that these would still oredicted as oreCopper and ingotCopper, these are free-form string assignments still defined by the modders in this theoretical system.
2) Mod B adds another copper ore, which is functionally identical, but also adds a crusher machine which can double the ore into dusts. The new dustCopper, since it is made in a non-vanilla machine, would NEED to be registered with oreCopper as a "parent" property by the modder MANUALLY in some way, and this is either assigned in the dustCopper item class itself (the dust implements a new interface called, say, ForgeResources), and/or the Crusher machine itself has some additional data which allows Forge to understand that all/some of it's products are part of this hierarchical resource system (e.g. Crusher implements ForgeResourcesProcessor). A child can be calculated automatically since the dusts add a valid furnace recipe. So now we have a simple relationship entry in the system, albeit with three members - a primary, secondary and tertiary resource - the new dustCopper inserted in the middle.
3) Mod C adds ANOTHER copper ore, still called oreCopper, but this one drops 2-4 dustCopper instead of itself. This block could have dustCopper as a "child" property, but this is also capable of being automatically set at the API level (the block has drops set, and it's drops are oredicted). However, this new  primary>secondary relationship conflicts with the Mod B oreCopper relationship; so.... heck I don't know, make it register as oreCopper2 and dustCopper2; because clearly they are NOT the same "value" copper ore as Mod A and Mod B above. The dustCopper from this mod should NOT inherit the same dustCopper recipes as Mod B. This "auto-isolation" could be configurable by the user, though.

So, in summary, Mod B's Ore and Ingots are completely skipped in registration - they're functionally identical to Mod A's ore and ingots. Or at least hidden and banned from being crafted/produced/whatever. As for ore gen though, well that's a WHOLE other story which I won't get into right now (is an oregen API akin to CoFH world generation outside the scope of Forge? Because, basically, that is what would work great - IF modders wanted to use the new system). And in the case of Mod C, well it's clearly not the same type of copper - the "value" of this new ore/dustCopper2 is not the same as original ore/dustCopper so it should not be treated as such.

Another two, much shorter, examples:

In a modpack I'm playing right now, I can process Mekanism's Osmium (the only mod in the pack that even has this ore) in MagnetiCraft's production line, which makes AOBD chunks/rubbles/pebbles/dusts (which is fine), dusts go into an Ender IO smeltery, and the output is.... a RotaryCraft Osmium ingot. What the? What business does RotaryCraft have taking over ingots from another mod when RotaryCraft itself has NO use or registration for an Osmium ore at all? In this case, the system described above would simply refuse RotaryCraft from registering it's Osmium ingot, and let the ACTUAL unique owner of ingotOsmium  - Mekanism - be it's only owner, as in no other ingots will even appear in the game. RotaryCraft's processing and output of Osmium should be Mekanism:ingotOsmium too. With the parent/child system, this is easily detected and enforced by Forge.

Last example, is a common thing unrelated to ores but to general ore dictionary entries. If the input items are NOT the same, should they be regarded as equivalent still? I can't think of any specific examples right now, but I think gears are a possible example IIRC. Railcraft gears are cheaper than Buildcraft gears, or something like that, but they are mostly interchangeable. Doesn't seem right that Forge allows this kind of thing. Or is this also beyond the scope of Forge, i.e. "let minetweaker scripter's sort it themselves"?

Phew, that was long. Sorry :)
